File: src/templates_py/templates_py/action_server_queue_goals.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from custom_interfaces.action import ExampleAction
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.executors import MultiThreadedExecutor
import time
import collections
import threading

class Server(Node):

    def __init__(self, node_name):
        super().__init__(node_name=node_name)

        self.goal_queue = collections.deque()
        self.goal_queue_lock = threading.Lock()
        self.current_goal = None

        self.action_server = ActionServer(
            node=self,
            action_type=ExampleAction,
            action_name='/example_action',
            execute_callback=self.execute_callback,
            callback_group=ReentrantCallbackGroup(),
            goal_callback=self.goal_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            cancel_callback=self.cancel_callback
        )

        self.get_logger().info(message='Action server is up')

    def execute_callback(self, goal_handle):
        try:
            self.get_logger().info(message='Executing new goal')
        
            goal = goal_handle.request.goal
            result = ExampleAction.Result()
            feedback = ExampleAction.Feedback()
            success = False
            cancel = False

            counter = 0
            max_num = 10

            while rclpy.ok():
                
                if goal_handle.is_cancel_requested:
                    cancel = True
                    break

                if counter == goal:
                    success = True
                    break

                if counter >= max_num:
                    break

                counter += 1

                feedback.feedback = counter
                goal_handle.publish_feedback(feedback)

                time.sleep(1.0)

            result.result = success

            if cancel:
                goal_handle.canceled()
                self.get_logger().warn(message='Goal canceled')
            elif success:
                goal_handle.succeed()
                self.get_logger().info(message='Goal succeeded')
            else:
                goal_handle.abort()
                self.get_logger().error(message='Goal aborted')

            return result
        
        finally:
            with self.goal_queue_lock:
                try:
                    self.current_goal = self.goal_queue.popleft()
                    self.get_logger().info(message='Next goal pulled from the queue')
                    self.current_goal.execute()
                except IndexError:
                    self.current_goal = None

    def goal_callback(self, goal_request):
        if goal_request.goal > 0:
            self.get_logger().info(message='Incoming goal accepted')
            return GoalResponse.ACCEPT
        else:
            self.get_logger().warn(message='Incoming goal rejected')
            return GoalResponse.REJECT
        
    def handle_accepted_callback(self, goal_handle):
        with self.goal_queue_lock:
            
            if self.current_goal is not None:
                self.goal_queue.append(goal_handle)
                self.get_logger().info(message='Goal put in the queue')
            else:
                self.current_goal = goal_handle
                self.current_goal.execute()


    def cancel_callback(self, goal_handle):
        self.get_logger().warn('Canceling goal')
        return CancelResponse.ACCEPT
    
def main(args=None):
    rclpy.init(args=args)
    action_server = Server(node_name='action_server')
    rclpy.spin(node=action_server, executor=MultiThreadedExecutor())
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/templates_cpp/src/action_server_queue_goals.cpp
#include <rclcpp/rclcpp.hpp>
#include <custom_interfaces/action/example_action.hpp>
#include <rclcpp_action/rclcpp_action.hpp>

typedef custom_interfaces::action::ExampleAction ExampleAction;
typedef rclcpp_action::ServerGoalHandle<ExampleAction> GoalHandle;

class Server : public rclcpp::Node
{
    public:

    Server(std::string node_name) : Node(node_name)
    {
        action_server = rclcpp_action::create_server<ExampleAction>
        (
            this->get_node_base_interface(),
            this->get_node_clock_interface(),
            this->get_node_logging_interface(),
            this->get_node_waitables_interface(),
            "/example_action",
            std::bind(&Server::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
            std::bind(&Server::handle_cancel, this, std::placeholders::_1),
            std::bind(&Server::handle_accepted, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "Action server is up");
    }

    private:

    void execute(const std::shared_ptr<GoalHandle> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Executing new goal");
        
        const auto goal = goal_handle->get_goal();
        auto result = std::make_shared<ExampleAction::Result>();
        auto feedback = std::make_shared<ExampleAction::Feedback>();
        bool success = false;
        bool cancel = false;

        int counter = 0;
        int max_num = 10;

        rclcpp::Rate rate(1.0);

        while (rclcpp::ok())
        {
            if (goal_handle->is_canceling())
            {
                cancel = true;
                break;
            }

            if (counter == goal->goal)
            {
                success = true;
                break;
            }

            if (counter >=max_num)
            {
                break;
            }

            counter++;

            feedback->feedback = counter;
            goal_handle->publish_feedback(feedback);

            rate.sleep();
        }

        result->result = success;

        if (cancel)
        {
            goal_handle->canceled(result);
            RCLCPP_WARN(this->get_logger(), "Goal canceled");
        }
        else if (success)
        {
            goal_handle->succeed(result);
            RCLCPP_INFO(this->get_logger(), "Goal succeeded");
        }
        else
        {
            goal_handle->abort(result);
            RCLCPP_ERROR(this->get_logger(), "Goal aborted");
        }

        goal_queue.pop();

        if (!goal_queue.empty())
        {
            current_goal = goal_queue.front();
            RCLCPP_INFO(this->get_logger(), "Next goal put from the queue");
            std::thread{std::bind(&Server::execute, this, std::placeholders::_1), current_goal}.detach();
        }
    }

    rclcpp_action::GoalResponse handle_goal
    (
        const rclcpp_action::GoalUUID &uuid,
        std::shared_ptr<const ExampleAction::Goal> goal_request
    )
    {
        (void)uuid;
        if (goal_request->goal > 0)
        {
            RCLCPP_INFO(this->get_logger(), "Incoming goal accepted");
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        }
        else
        {
            RCLCPP_WARN(this->get_logger(), "Incoming goal rejected");
            return rclcpp_action::GoalResponse::REJECT;
        }
    }

    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<GoalHandle> goal_handle)
    {
        RCLCPP_WARN(this->get_logger(), "Canceling goal");
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted(const std::shared_ptr<GoalHandle> goal_handle)
    {
        goal_queue.push(goal_handle);
        if (goal_queue.size() == 1)
        {
            current_goal = goal_queue.front();
            std::thread{std::bind(&Server::execute, this, std::placeholders::_1), current_goal}.detach();
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "Goal put in the queue");
        }
    }

    rclcpp_action::Server<ExampleAction>::SharedPtr action_server;
    std::queue<std::shared_ptr<GoalHandle>> goal_queue;
    std::shared_ptr<GoalHandle> current_goal;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto action_server = std::make_shared<Server>("action_server");
    rclcpp::spin(action_server);
    rclcpp::shutdown();

    return 0;
}